File: sammy/230302_study/프로_72412_순위 검색.py
def solution(info, query):
    answer = []
    data=[]
    
    # info split 후, 리스트에 저장
    for i in range(len(info)):
        data.append(list(info[i].split()))
    
    for q in query:
        query_split=q.split(' and ')
        food,score=query_split.pop().split()
        query_split.append(food)
        
        cnt=0
        for check in data:
            if int(check[-1])<int(score): continue
            
            for qu in query_split:
                if qu != '-' and qu not in check:
                    break
            else:
                cnt+=1
                    
        answer.append(cnt)
        
    return answer

print(solution(	["java backend junior pizza 150", "python frontend senior chicken 210", "python frontend senior chicken 150", "cpp backend senior pizza 260", "java backend junior chicken 80", "python backend senior chicken 50"], ["java and backend and junior and pizza 100", "python and frontend and senior and chicken 200", "cpp and - and senior and pizza 250", "- and backend and senior and - 150", "- and - and - and chicken 100", "- and - and - and - 150"]),	[1, 1, 1, 1, 2, 4])

File: James/230302_study/프로_72412_순위검색.java
import java.util.*;

class Solution {
    static Map<String, List<Integer>> map = new HashMap<>();
    
    public static void dfs(String[] data, String text, int cnt) {
        if(cnt == 4) {
            if(!map.containsKey(text)) {
                List<Integer> list = new ArrayList<>();
                map.put(text, list);
            }
            
            int score = Integer.parseInt(data[4]);
            map.get(text).add(score);
            return;
        }
        
        dfs(data, text + "-", cnt + 1);
        dfs(data, text + data[cnt], cnt + 1);
    }
    
    public static int binarySearch(String key, int score) {
        List<Integer> list = map.get(key);
        int start = 0;
        int end = list.size() - 1;
        
        while(start <= end) {
            int mid = (start + end) / 2;
            
            if(list.get(mid) < score) {
                start = mid + 1;
            }
            
            else {
                end = mid - 1;
            }
        }
        
        // 전체 점수 리스트의 갯수에서 score보다 작은 index(start)를 빼주면 score 이상의 점수 갯수가 됨. 
        return list.size() - start;
    }
    
    public int[] solution(String[] info, String[] query) {
        int[] answer = new int[query.length];
        
        for(int i = 0; i < info.length; i++) {
            String[] data = info[i].split(" ");
            dfs(data, "", 0);
        }
        
        // System.out.println(map.toString());
        
        for(String key : map.keySet()) {
            Collections.sort(map.get(key));
        }
        
        for(int i = 0; i < query.length; i++) {
            query[i] = query[i].replaceAll(" and ", "");
            String[] queryData = query[i].split(" ");
            
            if(map.containsKey(queryData[0])) {
                answer[i] = binarySearch(queryData[0], Integer.parseInt(queryData[1]));
            }
        }
    
        // System.out.println(Arrays.toString(query));
        
        return answer;
    }
}
